### ISSUE TITLE
Model: its `hasUnits()` method does different things depending on the signature that is used

### ISSUE THREAD

**Issue by agarny** [cellml/libcellml#719]
`Model::hasUnits()` has two signatures:
 1. `bool hasUnits(const std::string &name) const;`; and
 2. `bool hasUnits(const UnitsPtr &units) const;`.

There is clearly nothing wrong with that (on the contrary), but the problem is that the implementation is different whether we use the first or second signature. Indeed, although they both rely on a call to `Model::ModelImpl::findUnits()`, it happens that `findUnits()` also has two signatures:
 1. `std::vector<UnitsPtr>::iterator findUnits(const std::string &name);`; and
 2. `std::vector<UnitsPtr>::iterator findUnits(const UnitsPtr &units);`.

The problem is that they don't behave the same:
```c++
std::vector<UnitsPtr>::iterator Model::ModelImpl::findUnits(const std::string &name)
{
    return std::find_if(mUnits.begin(), mUnits.end(),
                        [=](const UnitsPtr &u) -> bool { return u->name() == name; });
}

std::vector<UnitsPtr>::iterator Model::ModelImpl::findUnits(const UnitsPtr &units)
{
    return std::find_if(mUnits.begin(), mUnits.end(),
                        [=](const UnitsPtr &u) -> bool { return units->name().empty() ? false : u->name() == units->name() && Units::equivalent(u, units); });
}
```

The bottom line is that I would expect a method that has different signatures to behave the same underneath. Anything else (like here) is a potential source for problems. (See https://github.com/cellml/libcellml/pull/716/files#r493927037 for some background on this issue.)

**Comment by hsorby** [comment]
I don't agree that methods with different signatures should behave the same underneath.  However the documentation should make it very clear about the differences between the methods.

**Comment by hsorby** [comment]
I think the underlying problem here is really an issue with using equivalence to match units.  I think that equality should be used here instead.

**Comment by agarny** [comment]
> I don't agree that methods with different signatures should behave the same underneath. However the documentation should make it very clear about the differences between the methods.

This makes things very confusing for the end-user (independent of the documentation). If we want to keep the current behaviour, then we really ought to rename the method.

**Comment by nickerso** [comment]
I agree with @hsorby - these are both find units methods and the different signature makes it very clear that the behaviour must be different. The switch from checking equivalence to equality would even ensure that these two different methods are behaving even more similar - you can just do a better comparison when you are given more information.

I think it is safe to close this one.

**Comment by agarny** [comment]
> these are both find units methods and the different signature makes it very clear that the behaviour must be different.

It's not because the signature is different that the behaviour should be different. For me, a different signature is a way to allow the user to use the same functionality in different ways.

Now, I appreciate that this method may not follow that logic (of mine, at least) since a `Units` object holds more information than a `std::string`. So, in turn, we might indeed expect a different behaviour unless the version with `UnitsPtr` was only to use the name of the units (i.e. as for the `std::string` version) and not the other information it holds.

I guess the bottom line is, as always, whatever the behaviour should match what the documentation says.
